From 5e3aff71f01f41254cdc7ecefc98a31be002dda0 Mon Sep 17 00:00:00 2001
From: "Dr. David Alan Gilbert" <dgilbert@redhat.com>
Date: Mon, 20 May 2019 12:17:36 +0100
Subject: [PATCH 18/29] DAX/unmap virtiofsd: Add wrappers for
 VHOST_USER_SLAVE_FS_IO

Add a wrapper to send VHOST_USER_SLAVE_FS_IO commands and a
further wrapper for sending a fuse_buf write using the FS_IO
slave command.

Signed-off-by: Dr. David Alan Gilbert <dgilbert@redhat.com>
---
 tools/virtiofsd/fuse_lowlevel.h | 24 +++++++++++++++++++++
 tools/virtiofsd/fuse_virtio.c   | 38 +++++++++++++++++++++++++++++++++
 2 files changed, 62 insertions(+)

diff --git a/tools/virtiofsd/fuse_lowlevel.h b/tools/virtiofsd/fuse_lowlevel.h
index 2259623776..866d122352 100644
--- a/tools/virtiofsd/fuse_lowlevel.h
+++ b/tools/virtiofsd/fuse_lowlevel.h
@@ -2022,4 +2022,28 @@ int fuse_virtio_unmap(struct fuse_session *se, VhostUserFSSlaveMsg *msg);
  */
 int fuse_virtio_sync(fuse_req_t req, VhostUserFSSlaveMsg *msg);
 
+/**
+ * For use with virtio-fs; request IO directly to memory
+ *
+ * @param se The current session
+ * @param msg A set of IO requests
+ * @param fd The fd to map
+ * @return Zero on success
+ */
+int fuse_virtio_io(struct fuse_session *se, VhostUserFSSlaveMsg *msg, int fd);
+
+/**
+ * For use with virtio-fs; wrapper for fuse_virtio_io for writes
+ * from memory to an fd
+ * @param req The request that triggered this action
+ * @param dst The destination (file) memory buffer
+ * @param dst_off Byte offset in the file
+ * @param src The source (memory) buffer
+ * @param src_off The GPA
+ * @param len Length in bytes
+ */
+ssize_t fuse_virtio_write(fuse_req_t req, const struct fuse_buf *dst,
+                          size_t dst_off, const struct fuse_buf *src,
+                          size_t src_off, size_t len);
+
 #endif /* FUSE_LOWLEVEL_H_ */
diff --git a/tools/virtiofsd/fuse_virtio.c b/tools/virtiofsd/fuse_virtio.c
index 24d9323665..abac0d0d2e 100644
--- a/tools/virtiofsd/fuse_virtio.c
+++ b/tools/virtiofsd/fuse_virtio.c
@@ -1071,3 +1071,41 @@ int fuse_virtio_sync(fuse_req_t req, VhostUserFSSlaveMsg *msg)
     return !vu_fs_cache_request(&req->se->virtio_dev->dev,
                                 VHOST_USER_SLAVE_FS_SYNC, -1, msg);
 }
+
+int fuse_virtio_io(struct fuse_session *se, VhostUserFSSlaveMsg *msg, int fd)
+{
+    if (!se->virtio_dev) {
+        return -ENODEV;
+    }
+    return !vu_fs_cache_request(&se->virtio_dev->dev,
+                                VHOST_USER_SLAVE_FS_IO, fd, msg);
+}
+
+/*
+ * Write to a file (dst) from an area of guest GPA (src) that probably
+ * isn't visible to the daemon.
+ */
+ssize_t fuse_virtio_write(fuse_req_t req, const struct fuse_buf *dst,
+                          size_t dst_off, const struct fuse_buf *src,
+                          size_t src_off, size_t len)
+{
+    VhostUserFSSlaveMsg msg = { 0 };
+
+    if (dst->flags & FUSE_BUF_FD_SEEK) {
+        msg.fd_offset[0] = dst->pos + dst_off;
+    } else {
+        off_t cur = lseek(dst->fd, 0, SEEK_CUR);
+        if (cur == (off_t)-1) {
+            return -errno;
+        }
+        msg.fd_offset[0] = cur;
+    }
+    msg.c_offset[0] = (uintptr_t)src->mem + src_off;
+    msg.len[0] = len;
+    msg.flags[0] = VHOST_USER_FS_FLAG_MAP_W;
+
+    bool result = !fuse_virtio_io(req->se, &msg, dst->fd);
+    /* TODO: Rework the result path to actually get length/error */
+    fuse_log(FUSE_LOG_DEBUG, "%s: result=%d\n", __func__, result);
+    return result ? len : -EIO;
+}
-- 
2.25.1

